src/architecture/serialization.md

```diff
@@ -1,533 +1,193 @@
+For example, `import "helpers.proto;"` can be used to get access to protobuf
+types `exonum.PublicKey`, `exonum.Hash`, etc. An example of such usage can be
+found in our [cryptocurrency example service][cryptocurrency].
+
+## Additional Validation for Protobuf-Generated Structures
+
+Protobuf is a versatile and flexible tool, which presents not only
+opportunities but also certain complications for the Exonum framework. For
+example, fields in protobuf cannot be fixed-size arrays, however, fixed-size
+arrays are required in Exonum (e.g. for hashes). It is possible to implement
+additional validations using the `.rs` protobuf-generated files. However, if
+users work with protobuf-generated structures, field validation would need to
+be performed every time they are used.
+
+To have validation performed only once for the whole structure, Exonum
+provides the conversion mechanism using the `ProtobufConvert` trait. This trait
+lets users automatically map
+their structures and the structures generated from `.proto` descriptions,
+providing a mechanism for validating protobuf-generated data. The structures
+for `ProtobufConvert` should have the same fields as the structures in
+`.proto` files, but can contain additional validation.
+
+The
+[`exonum_derive`](https://github.com/exonum/exonum/tree/master/exonum_derive)
+crate provides the ability to use structures typical for Exonum with all
+the required validations. So when using these structures users only need to
+implement `#[derive(ProtobufConvert)]` for them. If required, users can
+implement the [`ProtobufConvert`][convert] trait for any additional structures
+they need.
+
+For example, the protobuf description of the `TransactionRequest` message in
+Exonum is as follows:
+
+```protobuf
+message TransactionsRequest {
+  exonum.PublicKey to = 1;
+  repeated exonum.Hash txs = 2;
 }
 ```
 
-Here the message body is serialized as a `struct` with fields `first` and `second`
-having type `u64`.
-
-## Types to Be Supported in Future
-
-The current version does not support the serialization of the following types,
-but it is planned to be implemented in future:
-
-- [Enums][rust_enums]
-
-## Example
-
-Consider the structure with three fields:
-
-- pub_key: `PublicKey`  
-  99ace6c721db293b0ed5b487e6d6111f22a8c55d2a1b7606b6fa6e6c29671aa1
-
-- Owner: `String`  
-  Andrew
-
-- Balance: `u64`  
-  1234
-
-To serialize the structure, one may use macros like this:
+The corresponding `TransactionRequest` structure with `ProtobufConvert` has
+the following representation:
 
 ```rust
-#[macro_use] extern crate exonum;
-extern crate hex;
-use exonum::crypto::PublicKey;
-use exonum::storage::StorageValue;
-use hex::FromHex;
-
-encoding_struct! {
-    struct Wallet {
-        pub_key: &PublicKey,
-        owner: &str,
-        balance: u64,
-    }
+#[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Debug, ProtobufConvert)]
+#[exonum(pb = "proto::TransactionsRequest", crate = "crate")]
```

Comment:
@dvush I thought `crate = ...` part was made optional. Also, shouldn't it be `crate = "exonum"` if we're talking about *user*-defined structs? I think it would be confusing to provide an example of a structure defined in an `exonum` crate if the way it's defined differs from a way user-defined structs would be defined.

Reply:
Yes, it is optional. This one is copy from `exonum` crate, thats why it has this parameter.

This `crate = ""crate"` attribute should be removed.